src/bin/listbox-dnd.rs

```diff
@@ -0,0 +1,228 @@
+            let surface: cairo::ImageSurface =
+                cairo::ImageSurface::create(cairo::Format::ARgb32, alloc.width, alloc.height)
+                    .unwrap();
+            let surface_: &cairo::Surface = surface.as_ref();
+
+            {
+                ancestor.get_style_context().add_class("drag-icon");
+                let surface_ctx = cairo::Context::new(surface_);
+                ancestor.draw(&surface_ctx);
+                ancestor.get_style_context().remove_class("drag-icon");
+            }
+
+            let (x, y) = w.translate_coordinates(&ancestor, 0, 0).unwrap();
+            surface.set_device_offset(-x as f64, -y as f64);
+            ctx.drag_set_icon_surface(surface_);
+        }
+    }
+}
+
+fn drag_end<W: IsA<Widget>>(w: &W, _ctx: &DragContext) {
+    println!("drag_end({:?})", w);
+}
+
+fn drag_data_get<W: IsA<Widget>>(
+    widget_being_dragged: &mut Option<Widget>,
+    w: &W,
+    _ctx: &DragContext,
+    sel: &SelectionData,
+    _: u32,
+    _: u32,
+) {
+    println!("drag_data_get({:?}", w);
+
+    *widget_being_dragged = Some(w.as_ref().clone());
+    // We have to call this to emit drag_data_received() on the target
+    sel.set(&Atom::intern(""), 32, &[]);
+}
+
+fn drag_data_received<W: IsA<Widget>>(
+    widget_being_dragged: &Widget,
+    target: &W,
+    _ctx: &DragContext,
+    _x: i32,
+    _y: i32,
+    _sel: &SelectionData,
+    _info: u32,
+    _time: u32,
+) {
+    println!("drag_data_received({:?})", target);
+    println!("Widget being dragged: {:?}", widget_being_dragged);
+
+    // The widget being dragged is a EventBox because that's what we set as draggable. We need to
+    // get its containing ListBoxRow.
+    let source = match widget_being_dragged.get_ancestor(ListBoxRow::static_type()) {
+        None => {
+            println!("Can't get parent of source {:?}!!!", widget_being_dragged);
+            return;
+        }
+        Some(parent) => parent,
+    };
+    // Downcast it
+    let source = source.downcast_ref::<ListBoxRow>().unwrap();
+
+    // Same for the target
+    let target = match target.get_ancestor(ListBoxRow::static_type()) {
+        None => {
+            println!("Can't get parent of target {:?}!!!", widget_being_dragged);
+            return;
+        }
+        Some(parent) => parent,
+    };
+    let target = target.downcast_ref::<ListBoxRow>().unwrap();
+    let target_row = target.get_index();
+
+    if source == target {
+        println!("Source and target are the same. Aborting drag.");
+        return;
+    }
+
+    // Remove source from its parent
+    source
+        .get_parent()
+        .unwrap()
+        .downcast_ref::<Container>()
+        .unwrap()
+        .remove(source);
+
+    // Add it next to target
+    target
+        .get_parent()
+        .unwrap()
+        .downcast_ref::<ListBox>()
+        .unwrap()
+        .insert(source, target_row);
+}
+
+static CSS: &'static str = "
+    .drag-icon {
+        background: white;
```

Comment:
How this works with dark theme?
Maybe there some alias used for color for listbox background color?